### PR TITLE
📙[FEATURE] Ability to Restore workflow.

### PR DESCRIPTION
**Pull request by gokulprathin8**

### CODE REVIEW

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 1]
```diff
@@ -924,13 +908,27 @@ public ResponseEntity<Resource> fileGetter(ModelMap model, @PathVariable(value="
...
-			String history_id = BaseTool.isNull(request.getParameter("history_id")) ? 
+			String history_id = BaseTool.isNull(request.getParameter("history_id")) ?
 					new RandomString(18).nextString() : request.getParameter("history_id");
 			String[] hosts = request.getParameterValues("hosts[]");
 			String[] encrypted_password = request.getParameterValues("passwords[]");
 			String[] environments = request.getParameterValues("envs[]");
 			String[] passwords = RSAEncryptTool.getPasswords(encrypted_password, session.getId());
 			resp = wt.execute(history_id, id, mode, hosts, passwords, environments, token);
+
+            assert id != null;
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:919]: what will happen if id is null? why `wt.execute` won't fail if `id` is null here?

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 2]
```diff
@@ -924,13 +908,27 @@ public ResponseEntity<Resource> fileGetter(ModelMap model, @PathVariable(value="
...
+            Optional<Workflow> optionalWorkflow = workflowrepository.findById(id);
+			if (optionalWorkflow.isPresent()) {
+				Workflow workflow = optionalWorkflow.get();
+				Checkpoint checkpoint = new Checkpoint();
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:923]: I am not sure this is correct. Spring doesn't allow you to `new` an object and eventually save to the database.

File: src/main/resources/static/js/gw.history.js [hunk 3]
```diff
@@ -215,50 +215,74 @@ GW.history = {
 		"      <th scope=\"col\">Notes (Click to Edit)</th> "+
 		"      <th scope=\"col\">Status</th> "+
 		"      <th scope=\"col\">Action</th> "+
+        "      <th scope=\"col\">Checkpoint</th> "+
```
Comment (ZihengSun) [inline src/main/resources/static/js/gw.history.js:218]: I think I mentioned. This column should be removed.

Add a new button `restore` to the `Action` column.

File: src/main/resources/static/js/gw.history.js [hunk 4]
```diff
@@ -215,50 +215,74 @@ GW.history = {
...
+		"</div>" +
 		content;
-		
+
 		return content;
-		
+
 	},
 
+    getCheckpointTable: function(response, workflowId){
```
Comment (ZihengSun) [inline src/main/resources/static/js/gw.history.js:260]: No no. This is wrong understanding. Each workflow history is a checkpoint. The relationship between `history_id` and `checkpoint_id` is one to one mapping. There is no separate `checkpoint table`, only one table - history table.

File: src/main/resources/static/js/gw.workflow.js [hunk 5]
```diff
@@ -163,21 +163,25 @@ GW.workflow = {
 		"</div>"+
 		"<div id=\"main-workflow-info-history\" class=\"tabcontent-workflow generalshadow\" style=\"height:calc(100% - 265px);  overflow-y: scroll; left:0; margin:0; padding: 5px; display:none;\">"+
 		'   <div class="row" id="workflow-history-container" style="padding:0px;margin:0px; " >'+
-		
+
 	    '   </div>'+
 		"</div>"+
+		// "<div id=\"main-workflow-info-checkpoint\" class='tabcontent-workflow generalshadow' style='height: calc(100% - 265px); overflow-y: scroll; left:0; margin:0; padding: 5px; display:none;\'>"+
```
Comment (ZihengSun) [inline src/main/resources/static/js/gw.workflow.js:169]: just remove these commented out lines. They will not be used.

File: src/main/resources/static/js/gw.workflow.js [hunk 6]
```diff
@@ -1055,130 +1059,179 @@ GW.workflow = {
...
 
 			}
-			
+
+		});
+
+
+	},
+
+	checkpoint: function(workflowId) {
```
Comment (ZihengSun) [inline src/main/resources/static/js/gw.workflow.js:1147]: this function name should be `get_checkpoint` or `list_checkpoints_of_workflow`. Also, I am not sure this is necessary. Because we only need one restore button for each workflow history. That is it. No need to query checkpoints because workflow history already has it. 

In other words, we don't want `checkpoints` function. We only save checkpoints automatically when people run the workflow.

**Review by ZihengSun — CHANGES_REQUESTED** [review]
Back to you. There is no test for any of the new files. Please write some tests for them. Also, we need a separate PR to enable the unit testing CI/CD in Github actions.

File: src/main/java/com/gw/database/HistoryRepository.java [hunk 7]
```diff
@@ -131,6 +131,7 @@ public interface HistoryRepository extends JpaRepository<History, String> {
      * @param history_id The ID of the history record.
      * @return A list containing a history record and additional information about the associated process.
      */
-    @Query(value = "select * from history, gwprocess where history.history_id = ?1 and history.history_process=gwprocess.id", nativeQuery = true)
-    List<Object[]> findOneHistoryofProcess(String history_id);
```
[the remark this answers is not in the document]
Comment (gokulprathin8) [reply src/main/java/com/gw/database/HistoryRepository.java:135]: my ide showed no usage of the function, but i added it back to avoid any issues.

File: src/main/java/com/gw/jpa/Checkpoint.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.annotations.Type;
+
+import javax.persistence.*;
+import java.util.Date;
+import java.util.UUID;
+
+@Entity
+@Table(name = "WorkflowCheckpoint")
+public class Checkpoint {
```
[the remark this answers is not in the document]
Comment (gokulprathin8) [reply src/main/java/com/gw/jpa/Checkpoint.java:12]: yes, tested. works with old jar file

File: src/main/java/com/gw/web/GeoweaverController.java (hunk 1 above)
Comment (gokulprathin8) [reply src/main/java/com/gw/web/GeoweaverController.java:919]: it will throw a error if id is null, added it for testing. removing.

File: src/main/java/com/gw/web/GeoweaverController.java (hunk 2 above)
Comment (gokulprathin8) [reply src/main/java/com/gw/web/GeoweaverController.java:923]: we use `new` keyword for entity class to create a new object for that request. 
The new keyword is not used if it is a Bean or if the object is a singleton handled by spring boot in terms of creating the object only once, like for example database connection should be only created once when the application starts and never again. for such things we use dependency injection

in this case, we explicilty want the JPA to create new object to insert it into db and its common. we are doing the same for history and other places in the codebase.
image

File: src/main/resources/static/js/gw.history.js (hunk 4 above)
Comment (gokulprathin8) [reply src/main/resources/static/js/gw.history.js:260]: old code. removed.

File: src/main/resources/static/js/gw.history.js (hunk 3 above)
Comment (gokulprathin8) [reply src/main/resources/static/js/gw.history.js:218]: this line of code show up beside the action button like this:
image

File: src/main/java/com/gw/jpa/Checkpoint.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
+package com.gw.jpa;
+
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.annotations.Type;
+
+import javax.persistence.*;
```
Comment (ZihengSun) [inline src/main/java/com/gw/jpa/Checkpoint.java:6]: Don't use * @gokulprathin8

File: src/main/java/com/gw/tools/CheckpointTool.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+import com.gw.dto.checkpoint.CheckpointDTO;
+import com.gw.jpa.Checkpoint;
+import com.gw.jpa.GWProcess;
+import com.gw.jpa.History;
+import com.gw.jpa.Workflow;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
```
Comment (ZihengSun) [inline src/main/java/com/gw/tools/CheckpointTool.java:17]: Don't use * @gokulprathin8

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 11]
```diff
@@ -44,11 +33,7 @@
 import org.springframework.stereotype.Controller;
 import org.springframework.ui.Model;
 import org.springframework.ui.ModelMap;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestMethod;
-import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.*;
```
[the remark this answers is not in the document]
Comment (ZihengSun) [reply src/main/java/com/gw/web/GeoweaverController.java:51]: Don't use * @gokulprathin8

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 12]
```diff
@@ -57,65 +57,71 @@
...
-    
+
     @Autowired
     GWSearchTool st;
-    
+
     @Autowired
-    FileTool ft;
-    
+	FileTool ft;
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:87]: There must be some problem with your IDE. Why the space before `FileTool` is weird. @gokulprathin8

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 13]
```diff
@@ -57,65 +57,71 @@
...
+
     @Autowired
     SSHSession sshSession;
-    
+
     // Configuration property for upload file path.
     @Value("${geoweaver.upload_file_path}")
     String upload_file_path;
 
+	@Autowired
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:108]: Space problem.

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 14]
```diff
@@ -185,7 +191,6 @@ public void destroy() {
                     resp = hist.deleteById(id);
                     break;
                 case "clear_nodes_edges":
-                    assert id != null;
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:188]: why deleting this line?

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 15]
```diff
@@ -417,50 +422,50 @@ public String showResetPasswordForm(@Param(value = "token") String token, Model
...
 	public @ResponseBody String skip_process(ModelMap model, WebRequest request){
 
 		String resp = null;
-		
+
 		try {
-			
+
 			String workflow_id = request.getParameter("workflowid");
-
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:463]: There are so many space changes. Can you do these changes in a separate PR? It is very confusing and not easy to find the changes that really matter.

File: src/main/java/com/gw/web/GeoweaverController.java [hunk 16]
```diff
@@ -939,7 +950,7 @@ public ResponseEntity<Resource> fileGetter(ModelMap model, @PathVariable(value="
 		return resp;
 	}
 
-	
+
```
Comment (ZihengSun) [inline src/main/java/com/gw/web/GeoweaverController.java:953]: Why need two lines of space? What is the normal space between two functions?

File: src/main/resources/static/js/gw.history.js [hunk 17]
```diff
@@ -1,66 +1,66 @@
...
+ *
  */
 
 GW.history = {
 
     deleteAllJupyter: function(hostid, callback){
 
         if(confirm("WARNING: Are you sure to remove all the history? This is permanent and cannot be recovered.")){
-            
+
```
Comment (ZihengSun) [inline src/main/resources/static/js/gw.history.js:13]: you should separate these space changes into another PR. Don't pollute the real purpose of this PR with all these irrelevant changes.

**Review by ZihengSun — CHANGES_REQUESTED** [review]
Back to you

File: src/main/java/com/gw/web/GeoweaverController.java (hunk 14 above)
Comment (gokulprathin8) [reply src/main/java/com/gw/web/GeoweaverController.java:188]: added it back

**Review by ZihengSun — APPROVED** [review]
lg
